### DISCUSSION TITLE
Can't wrap my head around the DialService interface

### DISCUSSION THREAD

**Discussion by itsthesteve** [benbjohnson/wtf#87] [Q&A]
I'm getting into Go and ran across [this article](https://www.gobeyond.dev/packages-as-layers/). I've been combing through this repo and it's super helpful! The one thing I've been scratching my head about is the `DialService` interface mentioned in the article. I get the concept of interfaces, but I don't see where you're actually referencing it. There are other `DialService` structs with the receiver methods, but the interface is just... kinda there, and has the same name as the structs created elsewhere. I though there was some Go magic that does automagic application if an interface/struct have the same name but no. 

Example
```
type Something interface {
	DoSomething()
	DoAnotherThing()
}

type AnotherThing struct {
	Something // I would think this type of reference would be somewhere in the code
}

func (a *AnotherThing) DoSomething() {
	log.Println("Doing something")
}


x := AnotherThing{}
x.DoSomething()
x.DoAnotherThing() // invalid memory address as it's not implemented
```

I'm still learning this stuff but this stumped me. Thanks!

**Comment by vaidd4** [comment]
The idea in Go is that you don't have an `implements` keyword.
If you want your `type` to be compatible with an `interface`, simply implement each member of the interface for the type.
(In the standard library many interface just have one member to be easily implemented like https://pkg.go.dev/io#Reader)

Go by example is a nice ressource for that kind of things:  
https://gobyexample.com/interfaces

**Comment by mgrijalva** [comment]
This is where VSCode, Intellij, and other IDEs come in use. Just right-click the interface and pick "Go to Implementations". Otherwise, as noted above, you'd have to search for them yourself since interfaces are implemented implicitly in go, not explicitly (like with the `implements` keyword).

You can see it's implemented in 3 different places in the app:
![Screenshot 2024-06-12 at 2 26 42 PM](https://github.com/benbjohnson/wtf/assets/2575126/7f3aa3d9-ef05-4dfd-a7a1-ea19b3f61660)
